File: ios/GymTaxx/Utilities/GymWeek.swift
//
//  GymWeek.swift
//  GymTaxx
//

import Foundation

/// Week boundaries for one person's challenge. Weeks run Monday 00:00 to Sunday
/// 23:59 in *their* time zone.
///
/// A time zone belongs to a participation, not to the app. It is read from the
/// phone when someone joins and then stored, for the same reason the currency is:
/// a holiday or a work trip must never quietly move a deadline mid-challenge. A
/// New Yorker's Sunday ends at midnight in New York, five hours after London's.
///
/// Rows created before this existed have no stored zone and fall back to London,
/// which is exactly the schedule they have been running on — nobody's deadline
/// shifts underneath them.
nonisolated struct GymWeek: Sendable, Equatable {

    /// The zone every pre-1.1 challenge ran on, and the fallback for anything
    /// unrecognised.
    static let defaultZoneIdentifier = "Europe/London"

    static let londonZone: TimeZone =
        TimeZone(identifier: defaultZoneIdentifier) ?? TimeZone(secondsFromGMT: 0) ?? .gmt

    let timeZone: TimeZone

    /// Monday-first Gregorian calendar in this participation's zone.
    let calendar: Calendar

    init(timeZone: TimeZone) {
        self.timeZone = timeZone
        var calendar = Calendar(identifier: .gregorian)
        calendar.firstWeekday = 2 // Monday
        calendar.timeZone = timeZone
        self.calendar = calendar
    }

    /// Reads a stored zone, tolerating anything unexpected rather than failing —
    /// a bad identifier must never be the reason a week can't be calculated.
    init(storedIdentifier: String?) {
        self.init(timeZone: TimeZone(identifier: storedIdentifier ?? "") ?? Self.londonZone)
    }

    /// The zone for a *new* participation, from the phone. Read once at join time
    /// and then stored.
    static var forCurrentDevice: GymWeek { GymWeek(timeZone: .current) }

    /// London — the schedule of every challenge that started before personal time
    /// zones existed.
    static let london = GymWeek(timeZone: londonZone)

    /// The Monday 00:00 that opens the week containing `date`.
    func monday(of date: Date) -> Date {
        calendar.dateInterval(of: .weekOfYear, for: date)?.start
            ?? calendar.startOfDay(for: date)
    }

    /// Which challenge week `date` falls in, counting from the Monday that opened
    /// the challenge. A user who joins on a Thursday is in week 0 until Sunday
    /// midnight, so their first week is genuinely short — deliberate, because
    /// weeks are calendar weeks, not personal 7-day windows.
    func index(for date: Date, start: Date) -> Int {
        let from = monday(of: start)
        let to = monday(of: date)
        let days = calendar.dateComponents([.day], from: from, to: to).day ?? 0
        return max(0, days) / 7
    }

    /// The Sunday that closes the given challenge week.
    func lastDay(ofWeek index: Int, start: Date) -> Date {
        let opening = monday(of: start)
        let days = (index * 7) + 6
        return calendar.date(byAdding: .day, value: days, to: opening) ?? opening
    }

    /// The Monday that opens the given challenge week.
    func firstDay(ofWeek index: Int, start: Date) -> Date {
        let opening = monday(of: start)
        return calendar.date(byAdding: .day, value: index * 7, to: opening) ?? opening
    }

    /// The instant a challenge week's Sunday runs out — the deadline itself.
    ///
    /// Midnight that opens the following Monday, so "ends Sunday" and "counted up
    /// to Sunday midnight" are the same moment rather than one being a day out.
    func deadline(ofWeek index: Int, start: Date) -> Date {
        let opening = monday(of: start)
        let days = (index + 1) * 7
        return calendar.date(byAdding: .day, value: days, to: opening) ?? opening
    }

    /// The number of whole days between today and `date` in this zone.
    func daysUntil(_ date: Date, from now: Date = Date()) -> Int {
        let days = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: now),
            to: date
        ).day ?? 0
        return max(0, days)
    }

    /// Adds whole weeks, staying on the same wall-clock time even across a
    /// daylight-saving change.
    func adding(weeks: Int, to date: Date) -> Date {
        calendar.date(byAdding: .day, value: weeks * 7, to: date)
            ?? date.addingTimeInterval(Double(weeks) * 7 * 86_400)
    }

    // MARK: - Weekly starts

    /// The Monday a user joining at `date` begins on.
    ///
    /// A fresh challenge opens every Monday, so joining on a Monday starts that
    /// same day and joining any other day waits at most six. Starts stay pinned to
    /// Mondays rather than running from the moment of payment, because every week
    /// boundary in the app is a calendar week -- a personal Thursday-to-Thursday
    /// window would make "workouts this week" mean two different things.
    ///
    /// Replaces the old monthly cohorts: without pot-splitting, synchronised
    /// starts bought nothing and cost a late joiner up to five weeks of waiting.
    func weeklyStart(onOrAfter date: Date) -> Date {
        let opening = monday(of: date)
        // Already Monday: start today rather than pushing a full week out.
        if calendar.isDate(date, inSameDayAs: opening) { return opening }
        return calendar.date(byAdding: .day, value: 7, to: opening) ?? opening
    }
}

File: web/src/lib/gymweek.ts
/**
 * Week boundaries, mirroring `backend/functions/_shared/gymweek.ts` and
 * `GymWeek.swift`.
 *
 * Challenges open on Mondays at 00:00 in the participant's *own* time zone,
 * fixed when they join. Getting this wrong shifts a start date by hours, or by a
 * whole day for someone joining on a Sunday evening in California — which is
 * exactly when TikTok traffic peaks.
 *
 * These three implementations must stay in step.
 */

/** The zone every pre-1.1 challenge ran on, and the fallback for bad input. */
export const DEFAULT_ZONE = "Europe/London";

const DAY_MS = 86_400_000;

/** A stored zone we can safely hand to `Intl`. */
export function safeZone(zone: string | null | undefined): string {
  if (!zone) return DEFAULT_ZONE;
  try {
    new Intl.DateTimeFormat("en-US", { timeZone: zone });
    return zone;
  } catch {
    console.error("gymweek: unusable time zone, falling back to London", zone);
    return DEFAULT_ZONE;
  }
}

/** The browser's current zone, used to fix a new joiner's schedule. */
export function currentZone(): string {
  return safeZone(Intl.DateTimeFormat().resolvedOptions().timeZone);
}

/** How far `zone` is ahead of UTC at a given instant, in milliseconds. */
function zoneOffsetMs(at: Date, zone: string): number {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone: zone,
    hour12: false,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
  }).formatToParts(at);

  const value = (type: string): number => Number(parts.find((part) => part.type === type)?.value ?? "0");

  const asIfUTC = Date.UTC(
    value("year"),
    value("month") - 1,
    value("day"),
    value("hour") % 24,
    value("minute"),
    value("second"),
  );
  return asIfUTC - at.getTime();
}

/** The instant of 00:00 in `zone` on the given local calendar date. */
function zoneMidnight(year: number, month: number, day: number, zone: string): Date {
  const guess = Date.UTC(year, month - 1, day);
  return new Date(guess - zoneOffsetMs(new Date(guess), zone));
}

/**
 * The Monday a user joining at `at` begins on, in their own zone.
 * Joining on a Monday starts that same day; any other day waits at most six.
 */
export function weeklyStart(at: Date, zone: string = DEFAULT_ZONE): Date {
  const tz = safeZone(zone);
  const local = new Date(at.getTime() + zoneOffsetMs(at, tz));
  const weekday = local.getUTCDay(); // 0 Sun, 1 Mon
  const daysAhead = weekday === 1 ? 0 : (8 - weekday) % 7;

  const target = new Date(local.getTime() + daysAhead * DAY_MS);
  return zoneMidnight(target.getUTCFullYear(), target.getUTCMonth() + 1, target.getUTCDate(), tz);
}

/** `start` advanced by whole weeks, landing on the same local wall-clock time. */
export function addWeeks(start: Date, weeks: number, zone: string = DEFAULT_ZONE): Date {
  const tz = safeZone(zone);
  const shifted = new Date(start.getTime() + weeks * 7 * DAY_MS);
  const local = new Date(shifted.getTime() + zoneOffsetMs(shifted, tz));
  return zoneMidnight(local.getUTCFullYear(), local.getUTCMonth() + 1, local.getUTCDate(), tz);
}

/** Whole weeks between two instants, rounded. */
export function weeksBetween(start: Date, end: Date): number {
  return Math.round((end.getTime() - start.getTime()) / (7 * DAY_MS));
}

/**
 * Which week of the challenge `at` falls in, 1-based and clamped to the
 * challenge length so a late viewer never sees "week 5 of 4".
 */
export function weekIndex(start: Date, at: Date, totalWeeks: number): number {
  const elapsed = Math.floor((at.getTime() - start.getTime()) / (7 * DAY_MS));
  return Math.min(Math.max(elapsed + 1, 1), totalWeeks);
}

/** Start (inclusive) and end (exclusive) of the week `at` falls in. */
export function weekBounds(start: Date, at: Date, totalWeeks: number, zone: string): { from: Date; to: Date } {
  const index = weekIndex(start, at, totalWeeks);
  const from = addWeeks(start, index - 1, zone);
  return { from, to: addWeeks(start, index, zone) };
}

/** "Monday 17 August" — the start date as people actually read it. */
export function formatStartDate(date: Date, zone: string, locale: string): string {
  return new Intl.DateTimeFormat(locale, {
    timeZone: safeZone(zone),
    weekday: "long",
    day: "numeric",
    month: "long",
  }).format(date);
}
